src/Lean/Meta/Tactic/Simp/Main.lean

```diff
@@ -402,41 +380,400 @@ def simpForall (e : Expr) : SimpM Result := withParent e do
+  -/
+  simpHaveTelescopeAux (info : HaveTelescopeInfo) (fixed : Std.HashSet Nat) (used : Array Bool) (e : Expr) (i : Nat) (xs : Array Expr) : SimpM SimpHaveResult := do
+    if h : i < info.haveInfo.size then
+      let hinfo := info.haveInfo[i]
+      -- `x` and `val` are the fvar and value with respect to the local context.
+      let x := hinfo.decl.toExpr
+      let val := hinfo.decl.value true
+      -- Invariant: `v == val.abstract xs`.
+      let .letE n t v b true := e | unreachable!
+      -- Universe levels to use for each of the `have_*` theorems. It's the level of `val` and the level of the body.
+      let us := [hinfo.level, info.level]
+      if !used.getD i true then
+        /-
+        Unused `have`: do not add `x` to the local context then `simp` only the body.
+        -/
+        trace[Debug.Meta.Tactic.simp] "have telescope; unused {n} := {val}"
+        /-
+        Complication: Unused `have`s might only be transitively unused.
+        This means that `b.hasLooseBVar 0` might actually be true.
+        This matters because `t` and `v` appear in the proof term.
+        We use a dummy `x` for debugging purposes. (Recall that `Expr.abstract` skips non-fvar/mvars.)
+        -/
+        let x := Expr.const `_simp_let_unused_dummy []
+        let rb ← simpHaveTelescopeAux info fixed used b (i + 1) (xs.push x)
+        let expr := rb.expr.lowerLooseBVars 1 1
+        let exprType := rb.exprType.lowerLooseBVars 1 1
+        let exprInit := Expr.app (Expr.lam n t rb.exprInit .default) v
+        let exprResult := rb.exprResult.lowerLooseBVars 1 1
+        if rb.modified then
+          let proof := mkApp6 (mkConst ``have_unused_dep' us) t exprType v (mkLambda n .default t rb.exprInit) expr
+            (mkLambda n .default t rb.proof)
+          return { expr, exprType, exprInit, exprResult, proof, modified := true }
+        else
+          -- If not modified, this must have been a non-transitively unused `have`, so no need for `dep` form.
+          let proof := mkApp6 (mkConst ``have_unused' us) t exprType v expr expr
+            (mkApp2 (mkConst ``Eq.refl [info.level]) exprType expr)
+          return { expr, exprType, exprInit, exprResult, proof, modified := true }
+      else if fixed.contains i then
+        /-
+        Fixed `have` (like `CongrArgKind.fixed`): dsimp the value and simp the body.
+        The variable appears in the type of the body.
+        -/
+        let val' ← dsimp val
+        trace[Debug.Meta.Tactic.simp] "have telescope; fixed {n} := {val} => {val'}"
+        let vModified := val != val'
+        let v' := if vModified then val'.abstract xs else v
+        withExistingLocalDecls [hinfo.decl] <| withNewLemmas #[x] do
+          let rb ← simpHaveTelescopeAux info fixed used b (i + 1) (xs.push x)
+          let expr := Expr.app (Expr.lam n t rb.expr .default) v'
+          let exprType := Expr.app (Expr.lam n t rb.exprType .default) v'
```

Comment:
In the rest of the code, you use `mkLambda` and `mkApp`, why have you changed the style her?